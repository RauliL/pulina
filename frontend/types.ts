import { AnyAction, Dispatch, Store } from 'redux';

/**
 * Enumeration of different channel events.
 */
export enum ChannelEventType {
  /** User has joined the channel. */
  JOIN = 'join',
  /** User has left the channel. */
  PART = 'part',
  /** User has quit the chat. */
  QUIT = 'quit',
  /** User sent a message. */
  MESSAGE = 'message',
  /** An error occurred. */
  ERROR = 'error',
}

export interface ChannelEvent {
  /** Unique identifier of the event. */
  id: string;
  /** Timestamp when the event occurred. */
  timestamp: number;
  /** Type of the event. */
  type: ChannelEventType;
  /** Name of the channel. */
  channel: string;
}

export interface ChannelJoinEvent extends ChannelEvent {
  /** Type of the event. */
  type: ChannelEventType.JOIN;
  /** Nickname of the user who joined. */
  nick: string;
}

export interface ChannelPartEvent extends ChannelEvent {
  /** Type of the event. */
  type: ChannelEventType.PART;
  /** Nickname of the user who left. */
  nick: string;
}

export interface ChannelQuitEvent extends ChannelEvent {
  /** Type of the event. */
  type: ChannelEventType.QUIT;
  /** Nickname of the user who quit. */
  nick: string;
}

export interface ChannelMessageEvent extends ChannelEvent {
  /** Type of the event. */
  type: ChannelEventType.MESSAGE;
  /** Nickname of the user who sent the message. */
  nick: string;
  /** The message that was sent. */
  message: string;
  /** Whether the message should be highlighted or not. */
  isHighlight: boolean;
}

export interface ChannelErrorEvent extends ChannelEvent {
  /** Type of the event. */
  type: ChannelEventType.ERROR;
  /** The error message. */
  message: string;
}

export interface Channel {
  /** Name of the channel. */
  name: string;
  /** List of nicknames on the channel. */
  users: string[];
  /** List of events that occurred on the channel. */
  events: ChannelEvent[];
  /** Whether the channel has unread messages or not. */
  hasUnreadMessages: boolean;
  /** Whether the channel has unread highlight or not. */
  hasUnreadHighlight: boolean;
}

export interface PulinaState {
  /** Currently used nickname. */
  nick?: string;
  /** Nickname related error reported by the server. */
  nickError?: string;
  /** Channels which the user currently is on. */
  channels: { [name: string]: Channel };
  /** Name of the currently viewed channel. */
  currentChannel?: string;
  /** Whether channel list should be visible or not. */
  channelListVisible: boolean;
  /** Whether user list should be visible or not. */
  userListVisible: boolean;
}

export type PulinaAction = AnyAction;

export type PulinaDispatch = Dispatch<PulinaAction>;

export type PulinaStore = Store<PulinaState, PulinaAction>;
