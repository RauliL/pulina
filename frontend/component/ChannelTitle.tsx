import * as React from 'react';
import { Button, Navbar } from 'reactstrap';

import { Channel } from '../types/channel';

export interface Props {
  channel: Channel;
  onToggleChannelList: () => void;
  onToggleUserList: () => void;
}

const ChannelTitle: React.SFC<Props> = (props) => (
  <Navbar className="border-bottom">
    <span className="navbar-brand">{props.channel.name}</span>
    <div className="float-right d-lg-none">
      <Button
        outline={true}
        className="mr-2"
        onClick={props.onToggleChannelList}
      >
        <i className="fas fa-hashtag"/>
      </Button>
      <Button outline={true} onClick={props.onToggleUserList}>
        <i className="fas fa-user"/>
      </Button>
    </div>
  </Navbar>
);

export default ChannelTitle;
