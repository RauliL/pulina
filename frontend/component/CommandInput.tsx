import { startsWith, trim } from 'lodash';
import * as React from 'react';
import { Button, Col, Container, Form, Input, Row } from 'reactstrap';

import { ClientCommand } from '../../common/command';
import { isBlank, isValidNick } from '../../common/utils';

import { parseCommand } from '../command-parser';
import { Channel } from '../types/channel';

export interface StateProps {
  /** Nickname of the user itself. */
  nick?: string;
  /** List of nicknames on the current channel. */
  users?: string[];
}

export interface OwnProps {
  /** Currently viewed channel. */
  currentChannel: Channel;
  /** Callback for sending a command. */
  onCommand: <T extends ClientCommand> (command: T) => void;
  /** Callback for notifying about command errors. */
  onCommandError: (errorMessage: string) => void;
}

export type Props = StateProps & OwnProps;

export interface State {
  input: string;
}

export default class CommandInput extends React.Component<Props, State> {
  public constructor(props: Props) {
    super(props);

    this.state = { input: '' };
  }

  public componentDidMount() {
    const input = document.getElementById('command');

    if (input) {
      input.focus();
    }
  }

  public render() {
    return (
      <Form onSubmit={this.onSubmit} className="border-top">
        <Container>
          <Row>
            <Col lg={11} className="mt-2 mb-2">
              <Input
                id="command"
                value={this.state.input}
                onChange={this.onChangeInput}
                onKeyDown={this.onKeyDown}
              />
            </Col>
            <Col lg={1} className="mt-2 mb-2">
              <Button type="submit" block={true}>
                Send
              </Button>
            </Col>
          </Row>
        </Container>
      </Form>
    );
  }

  private onSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    const { input } = this.state;

    event.preventDefault();
    this.setState({ input: '' });

    // Ignore empty input.
    if (!input || isBlank(input)) {
      return;
    }

    // Attempt to parse the input into a command.
    parseCommand(this.props.currentChannel, trim(input))
      .then(this.props.onCommand)
      .catch(this.props.onCommandError);
  };

  private onChangeInput = (event: React.FormEvent<HTMLInputElement>) => {
    this.setState({ input: event.currentTarget.value });
  };

  private onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key !== 'Tab') {
      return;
    }

    event.preventDefault();

    if (this.props.users && isValidNick(this.state.input)) {
      const users = this.props.users.filter((nick) => (
        nick !== this.props.nick && startsWith(nick, this.state.input)
      ));

      if (users.length > 0) {
        this.setState({ input: `${users[0]}: ` });
      }
    }
  };
}
