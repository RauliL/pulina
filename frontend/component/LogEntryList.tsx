import * as React from 'react';
import { Container } from 'reactstrap';

import { LogEntry } from '../types/log';

import LogEntryListItem from './LogEntryListItem';

export interface Props {
  list: LogEntry[];
}

const style: React.CSSProperties = {
  overflowY: 'auto',
  fontFamily: '"Fira Mono", monospace',
};

export default class LogEntryList extends React.Component<Props> {
  public componentDidUpdate(prevProps: Props) {
    // If the number of events in the log has changed since last update, scroll
    // down the log so that the most recent one is always visible.
    if (this.props.list.length !== prevProps.list.length) {
      const container = document.getElementById('channel-log');

      if (container) {
        container.scroll({
          top: container.scrollHeight,
          behavior: 'smooth',
        });
      }
    }
  }

  public render() {
    return (
      <Container id="channel-log" fluid={true} style={style}>
        {this.props.list.map((entry) => (
          <LogEntryListItem
            key={entry.id}
            entry={entry}
          />
        ))}
      </Container>
    );
  }
}
