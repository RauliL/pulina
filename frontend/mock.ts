import { Channel } from './types/channel';

export const mockChannel: Channel = {
  name: '#test-channel',
  users: [],
  log: [],
  hasUnreadMessages: false,
  hasUnreadHighlight: false,
};
