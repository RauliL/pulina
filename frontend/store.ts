import { createStore } from 'redux';

import { reducer } from './reducer';
import { Action, State, Store } from './types/store';

export const initializeStore = (): Store => createStore<State, Action, {}, {}>(
  reducer,
  process.env.NODE_ENV === 'development' &&
  window &&
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__() :
  undefined,
);
