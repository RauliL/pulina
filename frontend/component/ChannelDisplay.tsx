import * as React from 'react';

import { ClientCommand } from '../../common/command';

import CommandInput from '../container/CommandInput';
import { Channel } from '../types/channel';

import ChannelTitle from './ChannelTitle';
import LogEntryList from './LogEntryList';

export interface Props {
  channel: Channel;
  onCommand: <T extends ClientCommand> (command: T) => void;
  onCommandError: (errorMessage: string) => void;
  onToggleChannelList: () => void;
  onToggleUserList: () => void;
}

const style = {
  display: 'grid',
  height: '100vh',
  gridTemplateColumns: '1fr',
  gridTemplateRows: 'auto 10fr auto',
  gridTemplateAreas: '"." "." "."',
};

const ChannelDisplay: React.SFC<Props> = (props) => (
  <div style={style}>
    <ChannelTitle
      channel={props.channel}
      onToggleChannelList={props.onToggleChannelList}
      onToggleUserList={props.onToggleUserList}
    />
    <LogEntryList list={props.channel.log}/>
    <CommandInput
      currentChannel={props.channel}
      onCommand={props.onCommand}
      onCommandError={props.onCommandError}
    />
  </div>
);

export default ChannelDisplay;
