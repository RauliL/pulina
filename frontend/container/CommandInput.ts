import { connect } from 'react-redux';

import CommandInput, { StateProps } from '../component/CommandInput';
import { State } from '../types/store';
import { getCurrentChannel } from '../utils';

const mapStateToProps = (state: State): StateProps => {
  const currentChannel = getCurrentChannel(state);
  let users: string[] | undefined;

  if (currentChannel) {
    users = currentChannel.users;
  }

  return {
    nick: state.nick,
    users
  };
};

export default connect(mapStateToProps)(CommandInput);
