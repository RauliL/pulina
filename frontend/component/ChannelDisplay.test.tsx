import { mount, shallow } from 'enzyme';
import { defaultTo, noop } from 'lodash';
import * as React from 'react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';

import CommandInput from '../container/CommandInput';
import { mockChannel } from '../mock';
import { State } from '../types/store';

import ChannelDisplay, { Props } from './ChannelDisplay';
import ChannelTitle from './ChannelTitle';
import LogEntryList from './LogEntryList';

describe('<ChannelDisplay/>', () => {
  const mockStore = createMockStore<State>();
  const renderComponent = (props: Partial<Props> = {}) => (
    <Provider store={mockStore()}>
      <ChannelDisplay
        channel={defaultTo(props.channel, mockChannel)}
        onCommand={defaultTo(props.onCommand, noop)}
        onCommandError={defaultTo(props.onCommandError, noop)}
        onToggleChannelList={defaultTo(props.onToggleChannelList, noop)}
        onToggleUserList={defaultTo(props.onToggleUserList, noop)}
      />
    </Provider>
  );
  const mountComponent = (props: Partial<Props> = {}) => mount(
    renderComponent(props),
  );

  it('should render without throwing exception', () => {
    expect(shallow(renderComponent())).toMatchSnapshot();
  });

  it('should render <ChannelTitle/>', () => {
    expect(mountComponent().find(ChannelTitle)).toHaveLength(1);
  });

  it('should render <LogEntryList/>', () => {
    expect(mountComponent().find(LogEntryList)).toHaveLength(1);
  });

  it('should render <CommandInput/>', () => {
    expect(mountComponent().find(CommandInput)).toHaveLength(1);
  });
});
