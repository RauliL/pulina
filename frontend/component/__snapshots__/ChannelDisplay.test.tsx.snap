// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelDisplay/> should render without throwing exception 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <ChannelDisplay
    channel={
      Object {
        "hasUnreadHighlight": false,
        "hasUnreadMessages": false,
        "log": Array [],
        "name": "#test-channel",
        "users": Array [],
      }
    }
    onCommand={[Function]}
    onCommandError={[Function]}
    onToggleChannelList={[Function]}
    onToggleUserList={[Function]}
  />
</ContextProvider>
`;
